overlord/devicestate/systems.go

```diff
@@ -176,3 +186,212 @@ func seededSystemFromModeenv() (*seededSystem, error) {
+			if nonEssentialPresence != "" {
+				kind = fmt.Sprintf("non-essential but %q",
+					nonEssentialPresence)
+			}
+		}
+		info, present, err := getInfo(name)
+		if err != nil {
+			return fmt.Errorf("cannot obtain %v snap information: %v", kind, err)
+		}
+		if !essential && !present && nonEssentialPresence == "optional" {
+			// non-essential snap which is declared as optionally
+			// present in the model
+			return nil
+		}
+		// grab those
+		logger.Noticef("%v snap: %v", kind, name)
+		if !present {
+			return fmt.Errorf("internal error: %v snap %q not present", kind, name)
+		}
+		if _, ok := modelSnaps[info.MountFile()]; ok {
+			// we've already seen this snap
+			return nil
+		}
+		// present locally
+		optsSnaps = append(optsSnaps, &seedwriter.OptionsSnap{
+			Path: info.MountFile(),
+		})
+		modelSnaps[info.MountFile()] = info
+		return nil
+	}
+
+	// snapd is implicitly needed
+	const snapdIsEssential = true
+	if err := getModelSnap("snapd", snapdIsEssential, ""); err != nil {
+		return nil, "", err
+	}
+	for _, sn := range model.EssentialSnaps() {
+		const essential = true
+		if err := getModelSnap(sn.SnapName(), essential, ""); err != nil {
+			return nil, "", err
+		}
+	}
+	for _, sn := range model.SnapsWithoutEssential() {
+		const essential = false
+		if err := getModelSnap(sn.SnapName(), essential, sn.Presence); err != nil {
+			return nil, "", err
+		}
+	}
+	if err := w.SetOptionsSnaps(optsSnaps); err != nil {
+		return nil, "", err
+	}
+
+	newFetcher := func(save func(asserts.Assertion) error) asserts.Fetcher {
+		fromDB := func(ref *asserts.Ref) (asserts.Assertion, error) {
+			return ref.Resolve(db.Find)
+		}
+		return asserts.NewFetcher(db, fromDB, save)
+	}
+	f, err := w.Start(db, newFetcher)
+	if err != nil {
+		return nil, "", err
+	}
+	// past this point the system directory is present
+
+	localSnaps, err := w.LocalSnaps()
+	if err != nil {
+		return nil, recoverySystemDir, err
+	}
+
+	for _, sn := range localSnaps {
+		info, ok := modelSnaps[sn.Path]
+		if !ok {
+			return nil, recoverySystemDir, fmt.Errorf("internal error: no snap info for %q", sn.Path)
+		}
+		// TODO: the side info derived here can be different from what
+		// we have in snap.Info, but getting it this way can be
+		// expensive as we need to compute the hash, try to find a
+		// better way
+		_, aRefs, err := seedwriter.DeriveSideInfo(sn.Path, f, db)
+		if err != nil {
+			if !asserts.IsNotFound(err) {
+				return nil, recoverySystemDir, err
+			} else if info.SnapID != "" {
+				// snap info from state must have come
+				// from the store, so it is unexpected
+				// if no assertions for it were found
+				return nil, recoverySystemDir, fmt.Errorf("internal error: no assertions for asserted snap with ID: %v", info.SnapID)
+			}
+		}
+		if err := w.SetInfo(sn, info); err != nil {
+			return nil, recoverySystemDir, err
+		}
+		sn.ARefs = aRefs
+	}
+
+	if err := w.InfoDerived(); err != nil {
+		return nil, recoverySystemDir, err
+	}
+
+	for {
```

Comment:
are we confident that this loop is necessary and won't run infinitely? I'm concerned because the code in seedwriter is difficult for me to reason about and we also are not following the description in the writer.go which says we need to call SetInfo on the snaps returned from SnapsToDownload. Note that here we also specifically error out if there are snaps that are returned from SnapsToDownload(), which leads me to believe that we expect SnapsToDownload to always return an empty list, which implies that we don't expect the state of seedwriter to change, meaning that I think we could just drop the `for` here and just run the code once and if complete is not true, have a separate error.

My main concern here is that we are not using these functions in the way that the package seems to describe them, and I don't trust myself to conclude that we are doing the right thing in this infinite loop to prevent running infinitely, and it also seems to me that we don't expect this to ever run more than one iteration implying that we could eliminate the possibility of an infinite loop by just checking at the end if we need to run again and if we do error out

Of course it is possible that I'm not following all of the state changes inside seedwriter properly and calling SnapsToDownload() and Downloaded() in a loop is still necessary ... even when there are no snaps to ever download ... (maybe this means that the names are not the best and should be something like `SnapsToResolve()` and `Resolved()` ?)

Reply:
this is the correct way of using the seedwriter (in practice we might not need the loop but is expected to be used like this, see the doc comment to Downloaded), OTOH no it would be a bug and quite a change in the seedwriter if it never returned true, each SnapsToDownload return a different set of potential snaps to download as needed and there's a finite sequence of such sets (see  type snapsToDownloadSet and related constants in seedwriter). If we want to be paranoid we could  limit the for loop to a number of times though.

Reply:
related to your comment about SetInfo, that's a problem in the diagram actually in seedwriter (so hard to keep diagrams up-to-date),  SetInfo can/should be called also between LocalSnaps and InfoDerived, the doc comments of those two say as much

Reply:
thanks for the explanation, this is fine now I see